library/src/cache/patch_manager.rs

```diff
@@ -0,0 +1,883 @@
+                None
+            }
+        }
+    }
+
+    fn save_patches_state(&self) -> Result<()> {
+        let path = self.root_dir.join(PATCHES_STATE_FILE_NAME);
+        disk_io::write(&self.patches_state, &path)
+    }
+
+    /// The directory where all patch artifacts are stored.
+    fn patches_dir(&self) -> PathBuf {
+        self.root_dir.join(PATCHES_DIR_NAME)
+    }
+
+    /// The directory where artifacts for the patch with the given number are stored.
+    fn patch_dir(&self, patch_number: usize) -> PathBuf {
+        self.patches_dir().join(patch_number.to_string())
+    }
+
+    /// The path to the runnable patch artifact with the given number. Runnable patch artifact files are
+    /// named <patch_number>.vmcode
+    fn patch_artifact_path(&self, patch_number: usize) -> PathBuf {
+        self.patch_dir(patch_number).join(PATCH_ARTIFACT_FILENAME)
+    }
+
+    fn patch_info_for_number(&self, patch_number: usize) -> PatchInfo {
+        PatchInfo {
+            path: self.patch_artifact_path(patch_number),
+            number: patch_number,
+        }
+    }
+
+    /// Checks that the patch with the given number:
+    ///   - Has an artifact on disk
+    ///   - That artifact on disk is the same size it was when it was installed
+    ///
+    /// Returns Ok if the patch is bootable, or an error if it is not.
+    fn validate_patch_is_bootable(&self, patch: &PatchMetadata) -> Result<()> {
+        let artifact_path = self.patch_artifact_path(patch.number);
+        if !Path::exists(&artifact_path) {
+            bail!(
+                "Patch {} does not exist at {}",
+                patch.number,
+                artifact_path.display()
+            );
+        }
+
+        let artifact_size_on_disk = std::fs::metadata(&artifact_path)?.len();
+        if artifact_size_on_disk != patch.size {
+            bail!(
+                "Patch {} has size {} on disk, but expected size {}",
+                patch.number,
+                artifact_size_on_disk,
+                patch.size
+            );
+        }
+
+        Ok(())
+    }
+
+    fn delete_patch_artifacts(&mut self, patch_number: usize) -> Result<()> {
+        info!("Deleting patch artifacts for patch {}", patch_number);
+
+        let patch_dir = self.patch_dir(patch_number);
+
+        std::fs::remove_dir_all(&patch_dir)
+            .with_context(|| format!("Failed to delete patch dir {}", &patch_dir.display()))
+    }
+
+    /// Attempts to use the last successfully booted patch as the next boot patch. If the last successfully
+    /// booted patch is not bootable or has the same number as the patch we're falling back from, we clear it.
+    fn try_fall_back_to_last_booted_patch(&mut self) {
+        if let Some(next_boot_patch) = self.patches_state.next_boot_patch {
+            // If we have a next_boot_patch that we're falling back from, delete its artifacts.
+            if let Err(e) = self.delete_patch_artifacts(next_boot_patch.number) {
```

Comment:
I guess I didn't expect this "cleanup the current patch" to be in this method. 🤷

Reply:
It wasn't originally, but this function is used as a "we couldn't boot from the patch we wanted to, see if we can boot from the last good one" operation, and cleaning up the one we're falling back from is a part of that. It could maybe have a better name, but I'm not sure what that would be atm